laspy/copc.py

```diff
@@ -0,0 +1,863 @@
+                    HttpRangeStream(source.url),
+                    offset,
+                    size,
+                )
+            )
+
+        # results = [future.result() for future in concurrent.futures.as_completed(jobs)]
+        # results.sort(key=lambda x: x[1])
+        # for group_bytes, _ in results:
+        #     cc = citer.next(len(group_bytes))
+        #     cc[:] = np.frombuffer(group_bytes, np.uint8)
+
+        # We don't use concurrent.futures.as_completed
+        # as we need to keep the order
+        citer = ChunkIter(np.frombuffer(out_compressed_bytes, np.uint8))
+        for future in jobs:
+            group_bytes = future.result()
+            cc = citer.next(len(group_bytes))
+            cc[:] = np.frombuffer(group_bytes, np.uint8)
+
+
+class CopcReader:
+    """
+    Class allowing to do queries over a `COPC`_ LAZ
+
+    In short, COPC files are LAZ 1.4 files organized in a particular way
+    (Octree) making it possible to do spatial queries
+    as well as queries with a level of details.
+
+    CopcReader **requires** the ``lazrz`` backend to work.
+
+    Optionaly, if ``requests`` is installed, CopcReader can handle
+    Copc files that are on a remote HTTP server
+
+    This class *only* reads COPC files, it does not support normal
+    LAS/LAZ files.
+
+    To create an instance of it you'll likely
+    want to use the :meth:`.CopcReader.open` constructor
+
+
+    .. _COPC: https://github.com/copcio/copcio.github.io
+    """
+
+    def __init__(
+        self,
+        stream,
+        http_num_threads: int = DEFAULT_HTTP_WORKERS_NUM,
+        close_fd: bool = True,
+    ):
+        """
+        Creates a CopcReader.
+
+        Parameters
+        ---------
+
+        stream: the stream from where data can be read.
+                It must have the following file object methods:
+                read, seek, tell
+
+        http_num_threads: int, optional, default num cpu * 5
+            Number of worker threads to do concurent HTTP requests,
+            ignored when reading non-HTTP file
+
+
+        close_fd: optional, default bool
+            Whether the stream/file object shall be closed, this only work
+            when using the CopcReader in a with statement.
+
+        """
+        if lazrs is None:
+            raise LazError("COPC support requires the 'lazrs' backend")
+        self.source = stream
+        self.close_fd = close_fd
+        self.http_num_threads = http_num_threads
+
+        # The idea here is to limit as much as possible the number of read
+        # to a slow external ressource like an HTTP endpoint
+        # by pre-fetching in memory what we need.
+        # COPC files _must_ be 1.4 LAS
+        header_bytes = stream.read(LAS_HEADERS_SIZE["1.4"])
+        offset_to_point_data = struct.unpack_from("<I", header_bytes, offset=96)[0]
+        rest = stream.read(offset_to_point_data - len(header_bytes))
+        header_source = io.BytesIO(header_bytes + rest)
```

Comment:
That should probably what the implementation of `LasHeader.read_from` should do by default